Mathlib/CategoryTheory/Sites/Coherent.lean

```diff
@@ -82,4 +83,148 @@ lemma isSheaf_coherent [Precoherent C] (P : Cᵒᵖ ⥤ Type w) :
+    (H : ∀ a, EffectiveEpiFamily (Y_n a) (π_n a)) :
+EffectiveEpiFamily (fun (c : Σ a, β a) => Y_n c.fst c.snd) (fun c => π_n c.fst c.snd ≫ π c.fst)
+    := by
+  rw [← Sieve.effectiveEpimorphic_family]
+  suffices h₂ : (Sieve.generate (Presieve.ofArrows (fun (⟨a, b⟩ : Σ _, β _) => Y_n a b)
+        (fun ⟨a,b⟩ => π_n a b ≫ π a))) ∈ GrothendieckTopology.sieves (coherentTopology C) X by
+    change Nonempty _
+    rw [← Sieve.forallYonedaIsSheaf_iff_colimit]
+    exact fun W => coherentTopology.isSheaf_yoneda_obj W _ h₂
+  let h' := h
+  rw [← Sieve.effectiveEpimorphic_family] at h'
+  let H' := H
+  conv at H' =>
+    intro a
+    rw [← Sieve.effectiveEpimorphic_family]
+  -- Show that a covering sieve is a colimit, which implies the original set of arrows is regular
+  -- epimorphic. We use the transitivity property of saturation
+  apply Coverage.saturate.transitive X (Sieve.generate (Presieve.ofArrows Y π))
+  · apply Coverage.saturate.of
+    use α, inferInstance, Y, π
+    simp only [true_and]
+    exact Iff.mp (Sieve.effectiveEpimorphic_family Y π) h'
+  · intro V f ⟨Y₁, h, g, ⟨hY, hf⟩⟩
+    rw [← hf, Sieve.pullback_comp]
+    apply (coherentTopology C).pullback_stable'
+    apply coherentTopology.mem_sieves_of_hasEffectiveEpiFamily
+    -- Need to show that the pullback of the family `π_n` to a given `Y i` is effective epimorphic
+    rcases hY with ⟨i⟩
+    use β i, inferInstance, Y_n i, π_n i
+    constructor
+    · exact H i
+    · intro b
+      use Y_n i b, (𝟙 _), π_n i b ≫ π i
+      constructor
+      · exact ⟨(⟨i, b⟩ : Σ (i : α), β i)⟩
+      · exact Category.id_comp (π_n i b ≫ π i)
+
+/--
+A sieve belongs to the coherent topology if and only if it contains a finite
+`EffectiveEpiFamily`
+-/
+theorem coherentTopology.mem_sieves_iff_hasEffectiveEpiFamily (S : Sieve X) :
+(S ∈ GrothendieckTopology.sieves (coherentTopology C) X) ↔
+    (∃ (α : Type) (_ : Fintype α) (Y : α → C) (π : (a : α) → (Y a ⟶ X)),
+        EffectiveEpiFamily Y π ∧ (∀ a : α, (S.arrows) (π a)) )  := by
+  constructor
+  · intro h
+    induction' h with Y T hS  Y Y R S _ _ a b
+    · rcases hS with ⟨a, h, Y', π, h'⟩
+      use a, h, Y', π
+      constructor
+      · tauto
+      · intro a'
+        cases' h' with h_left h_right
+        simp only [Sieve.generate_apply]
+        use Y' a', 𝟙 Y' a', π a'
+        constructor
+        · rw [h_left]
+          exact Presieve.ofArrows.mk a'
+        · apply Category.id_comp
+    · use Unit, Unit.fintype, fun _ => Y, fun _ => (𝟙 Y)
+      cases' S with arrows downward_closed
+      exact ⟨inferInstance, by simp only [Sieve.top_apply, forall_const]⟩
+    · rcases a with ⟨α, w, Y₁, π, ⟨h₁,h₂⟩⟩
+      have H  := fun a => b (h₂ a)
+      rw [Classical.skolem] at H
+      rcases H with ⟨β, H⟩
+      rw [Classical.skolem] at H
+      rcases H with ⟨_, H⟩
+      rw [Classical.skolem] at H
+      rcases H with ⟨Y_n, H⟩
+      rw [Classical.skolem] at H
+      rcases H with ⟨π_n, H⟩
```

Comment:
```suggestion
      choose β _ Y_n π_n H using fun a => b (h₂ a)
```

Reply:
Thanks! That one's very useful to know..